pkg/rollup/resolution.go

```diff
@@ -0,0 +1,182 @@
+// This file and its contents are licensed under the Apache License 2.0.
+// Please see the included NOTICE for copyright information and
+// LICENSE for a copy of the license.
+
+package rollup
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/timescale/promscale/pkg/log"
+	"github.com/timescale/promscale/pkg/pgxconn"
+)
+
+const (
+	originalSchema          = "prom_data"
+	upperLimit              = 5000 // Maximum samples allowed
+	assumedScrapeInterval   = time.Second * 30
+	refreshRollupResolution = time.Minute * 30
+)
+
+type rollupInfo struct {
+	schemaName string
+	resolution time.Duration
+}
+
+type QueryHelper struct {
+	conn                pgxconn.PgxConn
+	ctx                 context.Context
+	refreshMtx          sync.RWMutex
+	metricType          map[string]string // metric_name: metric_type
+	resolutionsASC      []rollupInfo      // {schemaName, resolution} in ascending order of resolution.
+	downsamplingEnabled bool
+}
+
+func NewQueryHelper(ctx context.Context, conn pgxconn.PgxConn) (*QueryHelper, error) {
+	helper := &QueryHelper{
+		ctx:  ctx,
+		conn: conn,
+	}
+	if err := helper.runRefreshRoutine(refreshRollupResolution); err != nil {
+		return nil, fmt.Errorf("refresh: %w", err)
+	}
+	return helper, nil
+}
+
+// DecideSchema returns the schema name of the rollups that should be used for querying.
+// The returned schema represents a downsampled resolution that should be an optimal
+// resolution for querying.
+//
+// If no rollups exists or if downsampling is disabled, "prom_data" is returned.
+func (h *QueryHelper) DecideSchema(min, max int64) string {
+	h.refreshMtx.RLock()
+	defer h.refreshMtx.RUnlock()
+
+	if !h.downsamplingEnabled || len(h.resolutionsASC) == 0 {
+		return originalSchema
+	}
+	estimateSamples := func(resolution time.Duration) int64 {
+		return int64(float64(max-min) / resolution.Seconds())
+	}
+
+	numRawSamples := estimateSamples(assumedScrapeInterval)
+	if numRawSamples < upperLimit {
+		return originalSchema
+	}
+
+	for _, info := range h.resolutionsASC {
+		samples := estimateSamples(info.resolution)
+		if samples < upperLimit {
+			// The first highest resolution that is below upper limit is our answer,
+			// since it provides the highest granularity at the expected samples.
+			return info.schemaName
+		}
+	}
+	// All rollups are above upper limit. Hence, send the schema of the lowest resolution
+	// as this is the best we can do.
+	lowestRollup := h.resolutionsASC[len(h.resolutionsASC)-1]
```

Comment:
If order is ascending I would expect that the highest resolution is the last.

Reply:
Sorry, but highest resolution is the one which has the least interval, since that will give maximum granularity. The last will be 1 hour (in example of 5mins, 15mins, 1hour) and hence 1 hour is of lowest resolution.

Reply:
I believe as it is now is counter-intuitive. I think we should first align on what resolution is. In our case we can probably define it as number of data points per time interval. Let' say we take 1h as a time for calculating resolution. So if we pick 5m granularity, that will have maximum resolution since we can fit 12 x 5000 data points in one hour. If we take 1h granularity it will be 5000 data points in 1 hour. That's why I believe that 5m is our highest resolution (not lowest) since resolution is a density of data points.
One potential solution could be to rename all `resolution` terms with interval. That will be more inline with current logic. It would also avoid confusion with terms b/c we are actually storing rollup intervals and not resolution